### DISCUSSION TITLE
A missing layout option: split-empty-space-evenly ?

### DISCUSSION THREAD

**Discussion by magthe** [YaLTeR/niri#883] [Ideas]
I tried out niri a few months ago, but didn't make the switch. Now I've come back and found the option `always-center-single-columns`. I'm using it together with `center-focus-column "never"` and liking it... though I've realised I don't _love_ it. I think the current behaviour favours leaving windows where they are, and I can appreciate that's something some people like, but personally I'd like a behaviour that favours distributing empty space evenly on both edges of the screen (which will make all columns visible when possible). I'll try to illustrate what I mean.

## Setup

I have a config where I cycle between column widths of 1/3, 1/2, and 2/3s of my screen. My layout contains

```
layout {
    center-focused-column "never"
    always-center-single-column
    ...
}
```

## Scenario α

1. open 1st window
2. open 2nd window

### Current behaviour

```
---------------------------------------------
|              | window0      |             |
|              |              |             |
|   empty      |              |   empty     |
|              |              |             |
|   1/3        |    1/3       |   1/3       |
|              |              |             |
---------------------------------------------
```

```
---------------------------------------------
|              | window0      | window1     |
|              |              |             |
|   empty      |              |             |
|              |              |             |
|   1/3        |    1/3       |     1/3     |
|              |              |             |
---------------------------------------------
```

All the empty space is on the left.

### Suggested behaviour

```
---------------------------------------------
|              | window0      |             |
|              |              |             |
|   empty      |              |   empty     |
|              |              |             |
|   1/3        |    1/3       |   1/3       |
|              |              |             |
---------------------------------------------
```

```
---------------------------------------------
|  e   | window0      | window1      |  e   |
|  m   |              |              |  m   |
|  p   |    1/3       |    1/3       |  p   |
|  t   |              |              |  t   |
|  y   |              |              |  y   |
| 1/6  |              |              | 1/6  |
---------------------------------------------
```

There's equal amounts of empty space outside the columns.

## Scenario β

1. open 1st window
2. open 2nd window
3. open 3rd window
4. resize 3rd window to 2/3 size
5. close 3rd window

### Current behaviour

```
---------------------------------------------
|              | window0      |             |
|              |              |             |
|   empty      |              |   empty     |
|              |              |             |
|   1/3        |    1/3       |   1/3       |
|              |              |             |
---------------------------------------------
```

```
---------------------------------------------
|              | window0      | window1     |
|              |              |             |
|   empty      |              |             |
|              |              |             |
|   1/3        |    1/3       |     1/3     |
|              |              |             |
---------------------------------------------
```

All the empty space is on the left.

```
---------------------------------------------
| window0      | window1      | window2     |
|              |              |             |
|              |              |             |
|              |              |             |
|   1/3        |    1/3       |     1/3     |
|              |              |             |
---------------------------------------------
```

```
---------------------------------------------
| window1      | window2                    |
|              |                            |
|              |                            |
|              |                            |
|   1/3        |    2/3                     |
|              |                            |
---------------------------------------------
```

```
---------------------------------------------
| window1      |                            |
|              |                            |
|              |    empty                   |
|              |                            |
|   1/3        |    2/3                     |
|              |                            |
---------------------------------------------
```

The column with window0 is outside the screen, and all the empty space is on the right.

### Suggested behaviour

```
---------------------------------------------
|              | window0      |             |
|              |              |             |
|   empty      |              |   empty     |
|              |              |             |
|   1/3        |    1/3       |   1/3       |
|              |              |             |
---------------------------------------------
```

```
---------------------------------------------
|  e   | window0      | window1      |  e   |
|  m   |              |              |  m   |
|  p   |    1/3       |    1/3       |  p   |
|  t   |              |              |  t   |
|  y   |              |              |  y   |
| 1/6  |              |              | 1/6  |
---------------------------------------------
```

There's equal amounts of empty space outside the columns.

```
---------------------------------------------
| window0      | window1      | window2     |
|              |              |             |
|              |              |             |
|              |              |             |
|   1/3        |    1/3       |     1/3     |
|              |              |             |
---------------------------------------------
```

```
---------------------------------------------
| window1      | window2                    |
|              |                            |
|              |                            |
|              |                            |
|   1/3        |    2/3                     |
|              |                            |
---------------------------------------------
```

```
---------------------------------------------
|  e   | window0      | window1      |  e   |
|  m   |              |              |  m   |
|  p   |    1/3       |    1/3       |  p   |
|  t   |              |              |  t   |
|  y   |              |              |  y   |
| 1/6  |              |              | 1/6  |
---------------------------------------------
```

There's equal amounts of empty space outside the columns, and both columns are visible.

## Consequences

- Centring the focused column doesn't always have any effect as there has to be enough non-empty space on both sides.

**Comment by YaLTeR** [comment]
Yeah, this makes sense. PaperWM works this way, too.

One of the problems however is how to do interactive resize nicely. Currently in always-centered mode, it will resize twice as fast and remain centered. In this case however I'm afraid we may have to make do with stationary resizing followed by windows moving into place.

**Reply by magthe** [reply]
I think that's more than fair...TBH I have a suspicion it'd be slightly confusing if the window I'm interactively resizing is moving around at the same time.

Anyway, I'm happy to hear you think  it makes sense... I'm eager to try it out when it lands in a release :pray:

**Reply by EdJoPaTo** [reply]
Personally I'm most interested in moving the columns to fill the screen when there are more columns to the left. Having them centered when there are less than screen width doesn't seem so important to me personally. So I think this has two parts: filling the screen when there are more columns (off-screen) than screen width and what happens when it's less than that (= centering the remaining ones).

For the first part I think this works already fine interactively with Mod + Right Mouse Hold. (Interestingly it applies the `struts` once I stop resizing so a resizing followed by windows moving into place logic already exists?)
This seems especially helpful on `close-window` and `switch-preset-column-width` (especially after `maximize-column`).

I would really like that behavior to know that more windows on the left exist. Like I always know that more are on the right when I close one in the middle.
The introduction of the overview lead me back to explore niri again and the struts are also helping with this awareness but simply filling the screen on window close/resize would be way easier to grasp.
(Having a strut the left only when there are more windows on the left would be neat too)

Regarding the second part, the interactive change when there is space left: Centering once the resizing is done seems like a good way of doing it. (Especially as its already part of the current logic with struts.) Keeping the current logic of resizing twice as fast and offsetting the other columns to the left/right seems also not that bad to understand for users? Both seem like a nice approach to me.

Also tried to put `center-visible-columns` after `close-window` / `switch-preset-column-width` binds but sadly only a single action per binding is allowed. So that doesn't work as a workaround for now either. Even better for the use-case of keeping the columns right aligned when there are more than screen space would be a `focus-column-first` then `focus-column-last` as that would make sure put the last one on the right side filling the screen. Doesn't work with the single binding limit either and would involve more animations, having a dedicated action for that would solve the animation issue tho.

**Comment by Plyply99** [comment]
Would really like to see this go through. Offset columns make my OCD go off lol.

![Screenshot from 2025-01-24 09-49-39](https://github.com/user-attachments/assets/b82c60b2-7abe-4d73-8809-cef9625c52ce)


Maybe Add the width of the columns and divide with screen resolution. Not a programmer so apologies for any nonsensical thoughts.

**Reply by magthe** [reply]
It would only kick in if `width_of_all_columns < screen_width`, and then the offset is `(screen_width - width_of_all_columns) / 2`.

(At least if we're talking about the same offsets :grin: )

**Comment by magthe** [comment]
Now, I'd never dream of saying I'm proficient in Rust, and I'm completely ignorant of the niri codebase, but if, on the off chance, I'd like to look at the code that'd need to chance for this, where should I start looking? (@YaLTeR ?)

**Reply by YaLTeR** [reply]
In src/layout/scrolling.rs

**Comment by baduhai** [comment]
I've tried using niri many times, but I keep coming back to plasma+karousel for this very feature. All other scrolling window management implementations I've tried work like this; hyprscroller, karousel, paperwm.

**Comment by YaLTeR** [comment]
Added `center-visible-columns` in defd4c5c4db8b22b2a5db468c0618c37b30e95dd.

**Reply by baduhai** [reply]
Guess I'm switching to niri next release. This and the overview?!?

@YaLTeR is cookin!

**Reply by Plyply99** [reply]
Thank you yes!
Beautiful centered columns.

**Reply by magthe** [reply]
That's an action that has to be triggered, right? There's not a setting like `always-center-visible-columns`?

**Reply by YaLTeR** [reply]
Correct

**Comment by magthe** [comment]
I've now installed the latest release, modified my keybindings, and used it for a day. I'd say it's a really nice complement to what I described above.

I still miss having my windows hug the screen's edges if the width of the columns exceed my screen width and otherwise divide the empty space equally on both sides.

**Comment by baduhai** [comment]
I tried my hand at adding this feature with [this pr](https://github.com/YaLTeR/niri/pull/2644), and overwall it's working well, I'm even using it daily. But I'm running into some race condition that I can't consistently reproduce, any help would be appreciated!

 > One of the problems however is how to do interactive resize nicely

The way I got around this, is by conly centering ince the interactive resize was over, it's functioning pretty well.
